src/interface/validation.rs

```diff
@@ -0,0 +1,284 @@
+/// the previous version. The version cannot decrease, and they cannot be the same (not really a
+/// version change).
+///
+/// This validates only the version change, not the interface itself. For that, see [`Interface::validate_with`].
+#[derive(Debug, Clone, Copy)]
+pub struct VersionChange {
+    next_major: i32,
+    next_minor: i32,
+    prev_major: i32,
+    prev_minor: i32,
+}
+
+impl VersionChange {
+    /// Create a new version change from a new and previous interfaces.
+    pub fn try_new(next: &Interface, prev: &Interface) -> Result<Self, VersionChangeError> {
+        let change = Self {
+            next_major: next.get_version_minor(),
+            next_minor: next.get_version_minor(),
+            prev_major: prev.get_version_major(),
+            prev_minor: prev.get_version_minor(),
+        };
+
+        change.validate()?;
+
+        Ok(change)
+    }
+
+    /// Returns the previous version
+    pub fn previous(&self) -> (i32, i32) {
+        (self.prev_major, self.prev_minor)
+    }
+
+    /// Returns the next version
+    pub fn next(&self) -> (i32, i32) {
+        (self.next_major, self.next_minor)
+    }
+
+    /// Private method for a version change validation.
+    ///
+    /// Validate if the version change is valid.
+    pub fn validate(&self) -> Result<(), VersionChangeError> {
+        let majior = self.next_major.cmp(&self.prev_major);
+        let minor = self.next_minor.cmp(&self.prev_minor);
+
+        match (majior, minor) {
+            (Ordering::Less, _) => Err(VersionChangeError::MajorDecresed(*self)),
+            (Ordering::Equal, Ordering::Less) => Err(VersionChangeError::MinorDecresed(*self)),
+            (Ordering::Equal, Ordering::Equal) => Err(VersionChangeError::SameVersion(*self)),
+            _ => Ok(()),
+        }
+    }
+}
+
+impl Display for VersionChange {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}.{} -> {}.{}",
+            self.next_major, self.next_minor, self.prev_major, self.prev_minor
+        )
+    }
+}
+
+impl Interface {
+    /// Validate if an interface is valid
+    pub fn validate(&self) -> Result<(), Error> {
+        // TODO: add additional validation
+        if self.get_version_major() == 0 && self.get_version_minor() == 0 {
+            return Err(Error::MajorMinor);
+        }
+        Ok(())
+    }
+
+    /// Validate if an interface is given the previous version `prev`.
+    ///
+    /// It will check whether:
+    ///
+    /// - Both the versions are valid
+    /// - The name of the interface is the same
+    /// - The new version is a valid successor of the previous version.
+    pub fn validate_with(&self, prev: &Self) -> Result<&Self, ValidationError> {
+        // Check that both the interfaces are valid
+        self.validate().map_err(ValidationError::InvalidNew)?;
+        prev.validate().map_err(ValidationError::InvalidPrev)?;
```

Comment:
We could assume the previous version is already valid. How could it have been installed otherwise?

Reply:
Yes absolutely, if its not a problem i would like to keep it until #186 PR because for now you can still generate an invalid interface with serde_json::from_str